### ISSUE TITLE
shamefully-hoist breaks package isolation in a monorepo

### ISSUE THREAD

**Issue by rsweeney21** [wxt-dev/wxt#286]
### Describe the bug

wxt adds shamefully-hoist=true to an .npmrc file in the folder of the wxt extension. However, this setting changes hoisting for the entire monorepo. After running pnpm install, pnpm shows the following warning:

```
Scope: all 15 workspace projects
✔ The modules directory at "/Users/robert/code/ditto/node_modules" will be removed and reinstalled from scratch. Proceed? (Y/n) · true
```

Now the node_modules at the root of the monorepo has all modules from all packages in the monorepo hoisted. This breaks one of the main benefits of using pnpm.

### To Reproduce

1. mkdir bug-repro && cd bug-repro
2. npm create svelte@latest my-app
3. pnpx wxt@latest init my-ext
4. Create a pnpm-workspace.yaml with my-app and my-ext in them 
5. npm install

### Expected behavior

node_modules in root shouldn't have modules from my-app and my-ext

### Environment

```
  System:
    OS: macOS 14.1.2
    CPU: (10) arm64 Apple M1 Max
    Memory: 15.90 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.17.1 - ~/.nvm/versions/node/v18.17.1/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 9.6.7 - ~/.nvm/versions/node/v18.17.1/bin/npm
    pnpm: 8.12.0 - ~/.nvm/versions/node/v18.17.1/bin/pnpm
    bun: 1.0.15 - ~/.bun/bin/bun
  Browsers:
    Chrome: 120.0.6099.109
    Safari: 17.1.2
  npmPackages:
    wxt: ^0.12.0 => 0.12.0 
```

### Additional context

I love wxt. It's incredible!

**Comment by aklinker1** [comment]
I also use WXT in a monorepo, but we have an electron app that requires `shamefully-hoist=true`, so this isn't something I've thought about much. I created this project based off Nuxt, which, at the time, also required setting this flag. Now though, it doesn't seem like the case (https://github.com/nuxt/nuxt/issues/14146).

So I can look into fixing this in WXT as well. Right now, there are 2 dependencies that prevent it from working:

- `webextension-polyfill`
- `@webext-core/match-patterns`

Unfortunately, I'm not an expert on PNPM and module resolution, so I don't know why only these two packages cause problems.

But if you list those two dependencies in your `package.json`, you can remove the `.npmrc` file. You can see this in a non-monorepo project:

```ts
$ pnpx wxt@latest init wxt-without-hoisting
$ cd wxt-without-hoisting
$ rm .npmrc
$ pnpm i
$ pnpm build
...
[vite]: Rollup failed to resolve import "webextension-polyfill" from "virtual:wxt-background?/Users/aklinker1/Development/local/wxt-without-hoisting/entrypoints/background.ts".
...

$ pnpm i webextension-polyfill
$ pnpm build
...
[vite]: Rollup failed to resolve import "@webext-core/match-patterns" from "virtual:wxt-background?/Users/aklinker1/Development/local/wxt-without-hoisting/entrypoints/background.ts".
...

$ pnpm i @webext-core/match-patterns
$ pnpm build
...
✔ Built extension
```

> Same errors for dev mode.

So I don't really know where to start with this...

Nuxt solved their problems by listing `vue` and `vue-router` in their package.json. I will probably do this with `webextension-polyfill`, but I don't know why `@webext-core/match-patterns` is causing any problems...

**Comment by aklinker1** [comment]
Here's an issue for vite that talks more about it: https://github.com/vitejs/vite/issues/324

And here's a issue that this was fixed in: https://github.com/preactjs/prefresh/issues/75

Both talk about talk about packages importing a subdependency not listed in the library's package.json, but I do list `@webext-core/match-patterns` inside WXT's `package.json`, and it has 0 dependencies, so it's not importing any sub-dependencies.

**Comment by aklinker1** [comment]
Actually, I know what the problem is now. 

```
[vite]: Rollup failed to resolve import "@webext-core/match-patterns" from "virtual:wxt-background?/Users/aklinker1/Development/local/wxt-without-hoisting/entrypoints/background.ts".
```

The error says it's coming from a virtual input. These inputs are shipped with WXT, but are not bundled until build time. This means they're effectively apart of your project, not WXT. That means they can't import any package not listed in your project's `package.json`.

Solution: Export any utilities from WXT and import them from WXT in these virtual entrypoints instead of from the 3rd party package directly.

This may also work for `webextension-polyfill`, and we might not have to add it to the `package.json` of the templates. That would be ideal, because eventuall, I want to remove the polyfill. If it's internal to WXT, I could remove it at anytime without making any breaking changes.

**Comment by aklinker1** [comment]
@rsweeney21 Can you try out `v0.12.2-alpha2`? I've got it working locally for some test projects.

```sh
pnpm i -D wxt@0.12.2-alpha2
```

**Comment by AndrewWalsh** [comment]
@aklinker1 I encountered this issue today using the package manager `pnpm` and monorepo tool `turbo`. I installed `0.12.2-alpha2` and it worked for me.

**Comment by aklinker1** [comment]
Ok, I'll merge and release this today then!

**Comment by aklinker1** [comment]
Released in [`v0.12.2`](https://github.com/wxt-dev/wxt/releases/tag/v0.12.2)

**Comment by aklinker1** [comment]
@tesths Can you open a new issue? Your problem is unrelated to this issue
